upstairs/src/active_jobs.rs

```diff
@@ -96,183 +119,83 @@ impl ActiveJobs {
-                break;
-            } else if (job.work.is_write() | job.work.is_repair())
-                && impacted_blocks.conflicts(&job.impacted_blocks)
-            {
-                // If this is a write or repair and it impacts the same blocks
-                // as something already active, create a dependency.
-                dep.push(*id);
-            }
-        }
+        let dep = self.block_to_active.check_range(impacted_blocks, false);
+        self.block_to_active
+            .insert_range(impacted_blocks, read_id, false);
         dep
     }
 
-    // Build the list of dependencies for a live repair job.  These are jobs that
-    // must finish before this repair job can move begin.  Because we need all
-    // three repair jobs to happen lock step, we have to prevent any IO from
-    // hitting the same extent, which means any IO going to our ImpactedBlocks
-    // (the whole extent) must finish first (or come after) our job.
-    pub fn deps_for_live_repair(
-        &self,
-        impacted_blocks: ImpactedBlocks,
-        close_id: u64,
-        _ddef: RegionDefinition,
+    /// Inserts a set of repair IDs into the dependency tracker.
+    ///
+    /// Note that jobs are only added to the dependency tracker; they are not
+    /// added to the list of active jobs (because this may be a future
+    /// reservation, without an allocated `DownstairsIO`).
+    pub fn deps_for_repair(
+        &mut self,
+        repair_ids: ExtentRepairIDs,
+        extent: u64,
     ) -> Vec<u64> {
-        let num_jobs = self.len();
-        let mut deps: Vec<u64> = Vec::with_capacity(num_jobs);
-
-        // Search backwards in the list of active jobs, stop at the
-        // last flush
-        for (id, job) in self.iter().rev() {
-            // We are finding dependencies based on impacted blocks.
-            // We may have reserved future job IDs for repair, and it's possible
-            // that this job we are finding dependencies for now is actually a
-            // future repair job we reserved.  If so, don't include ourself in
-            // our own list of dependencies.
-            if job.ds_id > close_id {
-                continue;
-            }
-            // If this operation impacts the same blocks as something
-            // already active, create a dependency.
-            if impacted_blocks.conflicts(&job.impacted_blocks) {
-                deps.push(*id);
-            }
-
-            // A flush job won't show impacted blocks. We can stop looking
-            // for dependencies beyond the last flush.
-            if job.work.is_flush() {
-                deps.push(*id);
-                break;
-            }
-        }
-
-        deps
+        let blocks = ImpactedBlocks::InclusiveRange(
+            ImpactedAddr {
+                extent_id: extent,
+                block: 0,
+            },
+            ImpactedAddr {
+                extent_id: extent,
+                block: u64::MAX,
```

Comment:
This is the trickiest part, and I spent a while convincing myself that it worked, so you've right to be skeptical.

I'm going to use `(e, b)` as shorthand for `ImpactedAddr::ImpactedAddr { extent_id: e, block: b }` in the explanation below, and Rust syntax for inclusive (`..=`) and exclusive (`..`) ranges.

There's an annoying constraint on the `BlockMap`, which is that it can't necessarily know `extent_size` (i.e. blocks per extent) in advance.  We have a ton of unit tests that run without a `RegionDefinition`, and I didn't want to change them all.

In this case, we insert a range covering blocks `0..u64::MAX`, but this is _within a single extent_.  Because the `BlockMap` doesn't know `extent_size`, it doesn't do any modular arithmetic; it operates entirely using lexicographical order.  This means that `ImpactedAddr { extent_id: 2, block: u64::MAX }` is considered _below_ `ImpactedAddr { extent_id: 3, block: 0 }`.

Unintuitively, this all ends up working fine for queries.  For example, let's start with a dependency that spans two extents with the job ID 0:
```
(3, 5)..(4, 4) => 0
```
If we now repair extent 3 with job ID 1, then we'll split this range into:
```
(3, 5)..(3, u64::MAX) => 1
(4, 0)..(4, 4) => 0
```
Reading and writing still works as intended in this configuration.  For example, if we read the inclusive range `(3, 5)..=(4, 0)`, it spans both segments and reads both dependencies.

The only way that reads could be misleading is if the _caller_ does something wrong: if `extent_limit = 8`, and the caller reads `(3, 15)..=(4, 0)`, then the results would include both ranges.  However, constructing `(3, 15)` with `extent_limit = 8` is already an undefined `ImpactedAddr`, so it's not obvious to me that `BlockMap` is necessarily incorrect in its response.

Reply:
Ah, yeah, that makes sense, thanks for explaining it.